[BE-2] refactor: JWT 발급과정 코드 리팩토링

xxbeann opened the pull request:
## #️⃣연관된 이슈

> ex) #56, #55, #16  

## 📝작업 내용

issueRefreshToken 메소드 JwtTokenProvider가 토큰을 저장하는 책임 분리 여부 검토

AuthTokenService 만료되는 시점 반환 수정

grantType 삭제 및 리팩토링 여부

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
>
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
  - 리프레시 토큰이 영속적으로 저장되어 관리됩니다.
  - 인증 오류 유형에 ‘유효하지 않은 리프레시 토큰’과 ‘유효하지 않은 서명’이 추가되어 세분화된 오류 메시지가 제공됩니다.
  - Swagger(OpenAPI) 기반 API 문서화가 도입되어 인증 및 산 관련 API에 대한 명확한 설명이 추가되었습니다.

- **버그 수정**
  - 인증 토큰 발급 시 액세스 토큰 만료 시간이 정확하게 계산되어 적용됩니다.
  - 리프레시 토큰 검증 로직이 강화되어 잘못된 토큰에 대해 적절한 예외가 발생합니다.

- **기타**
  - macOS 시스템 파일(.DS_Store)이 버전 관리에서 제외됩니다.
  - 인증 토큰 관련 필드가 정리되어 일부 불필요한 정보가 제거되었습니다.
  - 리프레시 토큰 발급 방식이 간소화되어 저장 로직이 변경되었습니다.
  - 인증 관련 예외 클래스명이 변경되고, 예외 처리 방식이 개선되었습니다.
  - Redis 설정이 프로필별로 분리되고 SSL 설정이 추가되어 보안이 강화되었습니다.
  - CORS 허용 대상에 다양한 로컬 네트워크 IP가 추가되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xxbeann committed:
feat: .DS_Store 추가

xxbeann committed:
refcator: grantType 삭제

xxbeann committed:
refactor: issueAuthToken이 만료되는 시점 반환

기존 만료시간을 반환하던 코드에서 현재시간에 만료시간을 더한 만료되는 시점을 반환하도록 리팩토링.

xxbeann committed:
refactor: JwtTokenProvider로부터 refreshToken저장 로직 분리

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/AuthTokenService.java:
```diff
@@ -36,6 +43,16 @@ public AuthToken reissue(String refreshToken) {
         return issueAuthToken(userId);
     }
 
+    @Transactional
+    public String saveRefreshToken(Long userId) {
+        String refreshToken = jwtTokenProvider.issueRefreshToken(userId);
+
+        RefreshToken token = new RefreshToken(userId, refreshToken, refreshTokenExpiredTime);
+        refreshTokenRepository.save(token);
+
+        return refreshToken;
+    }
```
새롭게 토큰을 발급하고 저장하는게 아닌, 이미 발급된 토큰을 저장하는 로직이 되야 합니다. 

그리고 해당 메소드를 호출하는 곳은 상위의 `OAuthService`가 아닌, authToken을 전반적으로 관리하는 이 서비스에서 호출이 이루어져야 될 것 같습니다

Woojae-Jeong reviewed (commented):
수고하셨습니다~ 코멘트 달았는데 반영 부탁드릴게요

Woojae-Jeong commented:
그리고 추가적으로 `JwtTokenProvider` 에서 `existRefreshTokenByUserId()` 와 `deleteRefreshTokenByUserId` 이 부분도 분리를 해야하는지 고민해주세요

xxbeann committed:
Merge branch 'dev' into refactor-#2

xxbeann commented:
확인했습니다!

xxbeann committed:
fix: conflict repair

xxbeann committed:
refactor: refreshToken 저장 로직 분리

xxbeann committed:
refactor: issueAuthToken 메서드 저장 로직 추가

xxbeann committed:
refactor: INVALID_TOKEN_EXPIRATION 명명 수정

기존 INVALID_TOKEN_EXPIRATION으로 명명된 시그니쳐 에러를 재명명하였습니다.
다시 명명된 오류명은 SIGNATURE_EXCEPTION과 같습니다.
오류명을 수정한 이유는 다음과 같습니다.

첫째, parseSignedClaims메서드 안에서 발생할 수 있는 오류는 토큰 만료시 발생하는 예외와, 토큰 서명이 위조되었거나 신뢰할 수 없을 때 발생하는 오류입니다.
이때, 첫번째 예외는 EXPIRED_TOKEN_EXCEPTION으로 그 예외 내용을 잘 담고 있으나, 두 번째인 INVALID_TOKEN_EXPIRATION은 유효하지 않다는 의미에서는 포괄적이나,
서명 위족 혹은 신뢰할 수 없다는 구체적인 의미를 나타낸다고 생각하지 않았습니다.

둘째, AuthTokenService 42번째 라인에서 커스텀 에러를 정의해야하는데, 이때 쓰이는 에러가 INVALID REFRESH TOKEN 에러입니다.
즉, 리프레시 토큰이 유효하지 않다는 의미입니다. 이때 기존의 에러명과 혼란이 있을 수 있다고 판단하여 에러명을 변경하였습니다.

xxbeann committed:
feat: InvalidRefreshToken 커스텀 에러 정의

xxbeann committed:
refactor: 커스텀 에러 적용

xxbeann committed:
refactor: 기존 이름 변경

userId가 사용되지 않는 메서드인걸 발견하고, 메서드명을 변경했습니다.

xxbeann committed:
fix: fix conflict

xxbeann commented:
확인하시고 이상 없으면 머지 부탁드립니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/JwtTokenProvider.java:
```diff
@@ -91,8 +84,8 @@ private Claims getClaimsFromToken(String token, TokenType tokenType) {
             throw new ExpiredTokenException();
-        } catch (SignatureException e) {
-            throw new InvalidTokenException();
+        } catch (io.jsonwebtoken.security.SignatureException e) {
+            throw new SignatureException();
```
`io.jsonwebtoken,security` 는 패키지 import를 통해서 가독성을 향상 시킬 수 있을 것 같네요.
해당 사항은`AppleJwtHanlder`에도 적용시키시면 됩니다!

xxbeann replied:
네 적용시키겠습니다!

xxbeann replied:
해당부분이 제가 정의한 커스텀에러 이름 SignatureException과 겹쳐서 저렇게 쓴거라 수정이 어려울 듯 합니다...

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/AuthTokenService.java:
```diff
@@ -1,41 +1,57 @@
         return issueAuthToken(userId);
     }
 
+    @Transactional
+    public void saveRefreshToken(Long userId, String refreshToken) {
+        RefreshToken token = new RefreshToken(userId, refreshToken, refreshTokenExpiredTime);
+        refreshTokenRepository.save(token);
+    }
```
책임 분리 잘 적용해 주셨네요 👍 


전에 말씀드렸다싶이 `JwtTokenProvider`는 토큰 발급과 토큰 파싱에 대해서만 책임을 가집니다. 
그렇다면`JwtTokenProvider.deleteRefreshTokenByUserId()` 메소드가 `JwtTokenProivder` 내부에 있는 게 적절한지 한번 확인해주세요!

xxbeann replied:
아하 그 뜻이었군요! 리팩토링 해보겠습니다 :)
상세한 리뷰 감사합니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/JwtTokenProvider.java:
```diff
@@ -77,7 +70,7 @@ public Long getUserIdFromRefreshToken(String refreshToken) {
-    public boolean existRefreshTokenByUserId(String refreshToken) {
+    public boolean isValidRefreshToken(String refreshToken) {
         return refreshTokenRepository.existsByRefreshToken(refreshToken);
     }
```
해당 메소드도 과연 `JwtTokenProvider` 의 책임일까요? 고민해 보시면 좋을 것 같습니다!

Woojae-Jeong reviewed (commented):
리뷰드린 사항에 대해서 잘 적용해주셨네요!
다만 아직 책임 분리가 명확히 되지 않은 것 같네요! 간단한 내용들이니 금방하실 수 있을 겁니다 👍

xxbeann committed:
Merge branch 'dev' into refactor-#2

xxbeann committed:
refactor: JwtTokenProvider에서 deleteRefreshTokenByUserId메소드 분리

JwtTokenProvider에서 deleteRefreshTokenByUserId메소드를 분리했습니다.
기존 메소드는 AuthTokenService에서 logout시 호출되던 메소드입니다.
이를 logout service 안으로 집어넣어 수정했습니다.

xxbeann committed:
refactor: JwtTokenProvider에서 isValidRefreshToken 책임 분리

JwtTokenProvider에서 isValidRefreshToken 책임 분리하였습니다.
AuthTokenService에서 호출하던 부분을 수정하였습니다.

xxbeann committed:
refactor: application.yml 과 application-jpa.yml에 redis 설정 중복 해결

application.yml 과 application-jpa.yml에 중복으로 되어 있던  redis 설정을 toekn profile로 옮겼습니다.
그 이유는 다음과 같습니다.
Redis는 JPA와 무관한 전역적인 설정이나, 현재 프로젝트에서 토큰 관련으로 사용되기 때문.
따라서 전역적인 application.yml 보다는 application-token이 더 적합하다 생각했습니다.
사실 그렇게 큰 차이는 없어 보입니다...

xxbeann committed:
refactor: application.yml 과 application-jpa.yml에 redis 설정 중복 해결

xxbeann committed:
chore:실행권한 추가

xxbeann committed:
feat: local profile 활성화 추가

xxbeann committed:
chore: swagger 의존성 추가

xxbeann committed:
chore: SwaggerConfig 추가

xxbeann committed:
chore:MountainConroller swagger annotation 추가

xxbeann committed:
chore:OAuthControlle Swagger Annotatioin 추가

xxbeann committed:
chore: CORS Origin 추가

xxbeann committed:
refactor:주소 앞 공백 제거
